### ISSUE TITLE
Realtime power, power consumption showing wrong readings

### ISSUE THREAD

**Issue by Pixelw** [georgezhao2010/midea_ac_lan#296]
### HA core version

2023.5.4

### Intergration version

0.3.19

### Device type and model

Air Conditioner 22040047

### Used App

MSmartHome

### The description of problem

![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/c78f1f41-75d8-49c5-bb17-c8b3739e2766)
The AC keeps running, after applying update 0.3.19, the real time power readings seams several times smaller. 
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/79543e7c-1f62-4f60-b9bc-0b04e5949358)
the total consumption reading also drop a lot.

**Comment by georgezhao2010** [comment]
Was it correct in the previous version?

Could you eable the debug log and show me a log containing ‘body type’: ‘c1’, which looks like the following

```
Received: {'header': 'aa20ac00000000000803', 'body': 'c12101440000000000000000000000000000000001df', 'message type': '03', 'body type': 'c1'}
```

**Comment by Pixelw** [comment]
> Was it correct in v0.3.18?

Yes, it was almost accurate compare to the power meter

> Could you enable the debug log and show me a log containing ‘body type’: ‘c1’, which looks like the following

Here's few of them: 

```
 Received: {'header': 'aa20ac00000000000803', 'body': 'c1210144000574850000000000000000008430000163', 'message type': '03', 'body type': 'c1'}

Received: {'header': 'aa20ac00000000000803', 'body': 'c121014400057487000000000000000000819000014c', 'message type': '03', 'body type': 'c1'}

Received: {'header': 'aa20ac00000000000803', 'body': 'c1210144000574870000000000000000007350000169', 'message type': '03', 'body type': 'c1'}

Received: {'header': 'aa20ac00000000000803', 'body': 'c12101440005748800000000000000000073100001e5', 'message type': '03', 'body type': 'c1'}
```

**Comment by georgezhao2010** [comment]
> > Was it correct in v0.3.18?
> 
> Yes, it was almost accurate compare to the power meter
> 
> > Could you enable the debug log and show me a log containing ‘body type’: ‘c1’, which looks like the following
> 
> Here's few of them:
> 
> ```
>  Received: {'header': 'aa20ac00000000000803', 'body': 'c1210144000574850000000000000000008430000163', 'message type': '03', 'body type': 'c1'}
> 
> Received: {'header': 'aa20ac00000000000803', 'body': 'c121014400057487000000000000000000819000014c', 'message type': '03', 'body type': 'c1'}
> 
> Received: {'header': 'aa20ac00000000000803', 'body': 'c1210144000574870000000000000000007350000169', 'message type': '03', 'body type': 'c1'}
> 
> Received: {'header': 'aa20ac00000000000803', 'body': 'c12101440005748800000000000000000073100001e5', 'message type': '03', 'body type': 'c1'}
> ```
> 
> Hope this helps

I found out 3 diffrent methods for analyzing the power consumption, each yielding different results. But I have no idea which ne would be the most accurate for your specific needs. I suppose it was 3?

result 1
```
current_energy_consumption = 0.0
realtime_power = 1.1516
total_energy_consumption = 617.36
```

result 2
```
current_energy_consumption = 0.0
realtime_power = 29.456
total_energy_consumption = 357.512
```

result 3
```
current_energy_consumption = 0.0
realtime_power = 1151.6
total_energy_consumption = 617.36
```

In version v0.3.19, you have the ability to customize the method for analyzing power consumption. Please refer to the [Customize](https://github.com/georgezhao2010/midea_ac_lan/blob/master/doc/AC.md#customize) for more details.

**Comment by Pixelw** [comment]
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/35a8f008-f4ed-4894-a751-e52f07b44252)
Ok, now it seems getting right, Thanks. But I still wondering why this update broke this feature, due to the compatbility with other device?

**Comment by georgezhao2010** [comment]
> ![image](https://user-images.githubusercontent.com/27913491/265978011-35a8f008-f4ed-4894-a751-e52f07b44252.png) Ok, now it seems getting right, Thanks. But I still wondering why this update broke this feature, due to the compatbility with other device?

Yeap, It's a compatbility update.

**Comment by Pixelw** [comment]
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/d5e22c44-71f1-4329-b24d-d1518eb40671)
here's another problem I found that the power consumption seem have some abnormal drop while AC running

**Comment by Pixelw** [comment]
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/7f4e17f1-4c9b-4206-b796-34c5da29bf24)
which causes energy panel show like this 😂

**Comment by Pixelw** [comment]
above results occured in 0.3.20

**Comment by Pixelw** [comment]
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/ffcae8cd-9892-48b9-830f-7628b747ba92)
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/f6906829-962d-4072-9860-b3e71da90ef6)
deeper blue bar is from a external power meter, before version 0.3.19 they were almost showing the same results.
if you need some log I can assist you. @georgezhao2010

**Comment by georgezhao2010** [comment]
If possible, please provide 24-hour continuous logs that contain changing energy data, especially when the energy data decreases instead of increasing. please compress it and send it to my email georgezhao75@gmail.com

**Comment by georgezhao2010** [comment]
> deeper blue bar is from a external power meter, before version 0.3.19 they were almost showing the same results. if you need some log I can assist you. @georgezhao2010

If possible, please provide 24-hour continuous logs that contain changing energy data, especially when the energy data decreases instead of increasing. please compress it and send it to my email [georgezhao75@gmail.com](mailto:georgezhao75@gmail.com)

**Comment by georgezhao2010** [comment]
> deeper blue bar is from a external power meter, before version 0.3.19 they were almost showing the same results. if you need some log I can assist you. @georgezhao2010

I got your email and thank you.

following the logs of your device, I found the "power analysis method" set to 2 is appropriate for your ac instead of 3.

**Comment by Pixelw** [comment]
But method 2 seem cause the problem I issued earlier - the power reading is many times smaller.
![image](https://github.com/georgezhao2010/midea_ac_lan/assets/27913491/d3a85c6f-7cf3-4356-a954-188ec1a4e7b0)

**Comment by georgezhao2010** [comment]
> But method 2 seem cause the problem I issued earlier

The ‘smaller’ you mentioned might be due to changes in the calculation method. You can observe the data for a time to see if it matches.

Anyway, the method 2 is the same as the lua script of your device.

```lua
if (messageBytes[3] == 0x44) then
    keyP["analysis_value"] =
        "total_power_consumption,total_operating_consumption,current_operating_consumption,current_time_power"
    keyP["total_power_consumption"] =
        bit.bor(bit.bor(bit.bor(bit.lshift(messageBytes[4], 32),
                                bit.lshift(messageBytes[5], 16)),
                        bit.lshift(messageBytes[6], 8)), messageBytes[7])
    keyP["total_operating_consumption"] =
        bit.bor(bit.bor(bit.bor(bit.lshift(messageBytes[8], 32),
                                bit.lshift(messageBytes[9], 16)),
                        bit.lshift(messageBytes[10], 8)),
                messageBytes[11])
    keyP["current_operating_consumption"] =
        bit.bor(bit.bor(bit.bor(bit.lshift(messageBytes[12], 32),
                                bit.lshift(messageBytes[13], 16)),
                        bit.lshift(messageBytes[14], 8)),
                messageBytes[15])
    keyP["current_time_power"] =
        bit.bor(bit.bor(bit.lshift(messageBytes[16], 16),
                        bit.lshift(messageBytes[17], 8)),
                messageBytes[18])
end
```


[T_0000_AC_22040047_2022040701.lua.zip](https://github.com/georgezhao2010/midea_ac_lan/files/12656244/T_0000_AC_22040047_2022040701.lua.zip)

**Comment by Pixelw** [comment]
https://github.com/georgezhao2010/midea_ac_lan/blob/01b406ffc24eb417a8171d9f0ae21924dc3df832/custom_components/midea_ac_lan/midea/devices/ac/message.py#L497-L501


after 0.3.19:

https://github.com/georgezhao2010/midea_ac_lan/blob/dd086f25922e3a210dc06865b75991b935d81da7/custom_components/midea_ac_lan/midea/devices/ac/message.py#L496-L503


Simular in power consumption calcuation.

Seems you remove the ```XC1MessageBody.parse_value``` function before calulating, did you handle that before these function calls? I can't verify that right now.

**Comment by georgezhao2010** [comment]
> The algorithm before 0.3.18 seems to be the method 3, which works fine in old versions.

Have you looked at the lua file attached in [previous comment](https://github.com/georgezhao2010/midea_ac_lan/issues/296#issuecomment-1724860350)? It's downloaded from midea server  and it's exclusive to your device.

Why do you firmly believe that the data before 0.3.18 was correct?

**Comment by georgezhao2010** [comment]
> The algorithm before 0.3.18 seems to be the method 3, which works fine in old versions.

refer https://github.com/georgezhao2010/midea_ac_lan/issues/114

**Comment by georgezhao2010** [comment]
> The algorithm before 0.3.18 seems to be the method 3, which works fine in old versions.

I understand what you’re saying. You mean that the current power value is too small for an air conditioner that is running. I think this might be an issue with the decimal point. I will continue to search for available information to find a more accurate calculation method.

**Comment by Pixelw** [comment]
>  Have you looked at the lua file attached in https://github.com/georgezhao2010/midea_ac_lan/issues/296#issuecomment-1724860350? It's downloaded from midea server and it's exclusive to your device. 

With my little knowledge of lua, the calculation port to Python looks correct.

> Why do you firmly believe that the data before 0.3.18 was correct?

I discover these readings by exploring the options of your intergretion. Surprisingly the power reading was only about 3% smaller than that external power meter, so I think it's reasonable.

> Are the energy consumption values showing in your Meiju App?

No, I didn't found a page showing that.

**Comment by georgezhao2010** [comment]
> > Have you looked at the lua file attached in [#296 (comment)](https://github.com/georgezhao2010/midea_ac_lan/issues/296#issuecomment-1724860350)? It's downloaded from midea server and it's exclusive to your device.
> 
> With my little knowledge of lua, the calculation port to Python looks correct.
> 
> > Why do you firmly believe that the data before 0.3.18 was correct?
> 
> I discover these readings by exploring the options of your intergretion. Surprisingly the power reading was only about 3% smaller than that external power meter, so I think it's reasonable.
> 
> > Are the energy consumption values showing in your Meiju App?
> 
> No, I didn't found a page showing that.

Fine. If your basis comes from an external power meter, then you are correct.

So the calculation in LUA is wrong, and no one cares about that because the result is shown nowhere.

I'll add back `XC1MessageBody.parse_value` into method 1.

**Comment by georgezhao2010** [comment]
Fixed in v0.3.21

Set the power analysis method to 1.

**Comment by Pixelw** [comment]
Thank you!!
